File: compiler/packages/babel-plugin-react-compiler/src/Utils/DisjointSet.ts
/**
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {CompilerError} from '../CompilerError';

// Represents items which form disjoint sets.
export default class DisjointSet<T> {
  #entries: Map<T, T> = new Map();

  /*
   * Updates the graph to reflect that the given @param items form a set,
   * linking any previous sets that the items were part of into a single
   * set.
   */
  union(items: Array<T>): void {
    const first = items.shift();
    CompilerError.invariant(first != null, {
      reason: 'Expected set to be non-empty',
      description: null,
      loc: null,
      suggestions: null,
    });
    /*
     * determine an arbitrary "root" for this set: if the first
     * item already has a root then use that, otherwise the first item
     * will be the new root.
     */
    let root = this.find(first);
    if (root == null) {
      root = first;
      this.#entries.set(first, first);
    }
    // update remaining items (which may already be part of other sets)
    for (const item of items) {
      let itemParent = this.#entries.get(item);
      if (itemParent == null) {
        // new item, no existing set to update
        this.#entries.set(item, root);
        continue;
      } else if (itemParent === root) {
        continue;
      } else {
        let current = item;
        while (itemParent !== root) {
          this.#entries.set(current, root);
          current = itemParent;
          itemParent = this.#entries.get(current)!;
        }
      }
    }
  }

  /*
   * Finds the set to which the given @param item is associated, if @param item
   * is present in this set. If item is not present, returns null.
   *
   * Note that the returned value may be any item in the set to which the input
   * belongs: the only guarantee is that all items in a set will return the same
   * value in between calls to `union()`.
   */
  find(item: T): T | null {
    if (!this.#entries.has(item)) {
      return null;
    }
    const parent = this.#entries.get(item)!;
    if (parent === item) {
      // this is the root element
      return item;
    }
    // Recurse to find the root (caching all elements along the path to the root)
    const root = this.find(parent)!;
    // Cache the element itself
    this.#entries.set(item, root);
    return root;
  }

  /*
   * Forces the set into canonical form, ie with all items pointing directly to
   * their root, and returns a Map representing the mapping of items to their roots.
   */
  canonicalize(): Map<T, T> {
    const entries = new Map<T, T>();
    for (const item of this.#entries.keys()) {
      const root = this.find(item)!;
      entries.set(item, root);
    }
    return entries;
  }

  /*
   * Calls the provided callback once for each item in the disjoint set,
   * passing the @param item and the @param group to which it belongs.
   */
  forEach(fn: (item: T, group: T) => void): void {
    for (const item of this.#entries.keys()) {
      const group = this.find(item)!;
      fn(item, group);
    }
  }

  buildSets(): Array<Set<T>> {
    const ids: Map<T, number> = new Map();
    const sets: Map<number, Set<T>> = new Map();

    this.forEach((identifier, groupIdentifier) => {
      let id = ids.get(groupIdentifier);
      if (id == null) {
        id = ids.size;
        ids.set(groupIdentifier, id);
      }

      let set = sets.get(id);
      if (set === undefined) {
        set = new Set();
        sets.set(id, set);
      }
      set.add(identifier);
    });

    return [...sets.values()];
  }

  get size(): number {
    return this.#entries.size;
  }
}

File: compiler/crates/react_compiler_utils/src/disjoint_set.rs
// Copyright (c) Meta Platforms, Inc. and affiliates.
//
// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

//! A generic disjoint-set (union-find) data structure.
//!
//! Ported from TypeScript `src/Utils/DisjointSet.ts`.

use std::collections::HashSet;
use std::hash::Hash;

use indexmap::IndexMap;

/// A Union-Find data structure for grouping items into disjoint sets.
///
/// Corresponds to TS `DisjointSet<T>` in `src/Utils/DisjointSet.ts`.
/// Uses `IndexMap` to preserve insertion order (matching TS `Map` behavior).
pub struct DisjointSet<K: Copy + Eq + Hash> {
    entries: IndexMap<K, K>,
}

impl<K: Copy + Eq + Hash> DisjointSet<K> {
    pub fn new() -> Self {
        DisjointSet {
            entries: IndexMap::new(),
        }
    }

    /// Updates the graph to reflect that the given items form a set,
    /// linking any previous sets that the items were part of into a single set.
    ///
    /// Corresponds to TS `union(items: Array<T>): void`.
    pub fn union(&mut self, items: &[K]) {
        if items.is_empty() {
            return;
        }
        let root = self.find(items[0]);
        for &item in &items[1..] {
            let item_root = self.find(item);
            if item_root != root {
                self.entries.insert(item_root, root);
            }
        }
    }

    /// Find the root of the set containing `item`, with path compression.
    /// If `item` is not in the set, it is inserted as its own root.
    ///
    /// Note: callers that need null/None semantics for missing items should
    /// use `find_opt()` instead.
    pub fn find(&mut self, item: K) -> K {
        let parent = match self.entries.get(&item) {
            Some(&p) => p,
            None => {
                self.entries.insert(item, item);
                return item;
            }
        };
        if parent == item {
            return item;
        }
        let root = self.find(parent);
        self.entries.insert(item, root);
        root
    }

    /// Find the root of the set containing `item`, returning `None` if the item
    /// was never added to the set.
    ///
    /// Corresponds to TS `find(item: T): T | null`.
    pub fn find_opt(&mut self, item: K) -> Option<K> {
        if !self.entries.contains_key(&item) {
            return None;
        }
        Some(self.find(item))
    }

    /// Returns true if the item is present in the set.
    ///
    /// Corresponds to TS `has(item: T): boolean`.
    pub fn has(&self, item: K) -> bool {
        self.entries.contains_key(&item)
    }

    /// Forces the set into canonical form (all items pointing directly to their
    /// root) and returns a map of items to their roots.
    ///
    /// Corresponds to TS `canonicalize(): Map<T, T>`.
    pub fn canonicalize(&mut self) -> IndexMap<K, K> {
        let mut result = IndexMap::new();
        let keys: Vec<K> = self.entries.keys().copied().collect();
        for item in keys {
            let root = self.find(item);
            result.insert(item, root);
        }
        result
    }

    /// Calls the provided callback once for each item in the disjoint set,
    /// passing the item and the group root to which it belongs.
    ///
    /// Corresponds to TS `forEach(fn: (item: T, group: T) => void): void`.
    pub fn for_each<F>(&mut self, mut f: F)
    where
        F: FnMut(K, K),
    {
        let keys: Vec<K> = self.entries.keys().copied().collect();
        for item in keys {
            let group = self.find(item);
            f(item, group);
        }
    }

    /// Groups all items by their root and returns the groups as a list of sets.
    ///
    /// Corresponds to TS `buildSets(): Array<Set<T>>`.
    pub fn build_sets(&mut self) -> Vec<HashSet<K>> {
        let mut group_to_index: IndexMap<K, usize> = IndexMap::new();
        let mut sets: Vec<HashSet<K>> = Vec::new();
        let keys: Vec<K> = self.entries.keys().copied().collect();
        for item in keys {
            let group = self.find(item);
            let idx = match group_to_index.get(&group) {
                Some(&idx) => idx,
                None => {
                    let idx = sets.len();
                    group_to_index.insert(group, idx);
                    sets.push(HashSet::new());
                    idx
                }
            };
            sets[idx].insert(item);
        }
        sets
    }

    /// Returns the number of items in the set.
    ///
    /// Corresponds to TS `get size(): number`.
    pub fn len(&self) -> usize {
        self.entries.len()
    }

    pub fn is_empty(&self) -> bool {
        self.entries.is_empty()
    }
}
